Mode whitelist

lrusak opened the pull request:
@FernetMenta wanted me to PR this to get some more visibility. I'll add what I had already posted to the internal boards.

---

Just wanted to start a discussion that won't get lost in slack.

For those that don't know the display and refresh rate switching logic in kodi is rather complex yet it doesn't quite fit everyone's needs. @FernetMenta proposed that we create a set of `whitelisted` resolutions that one can select and create a logic for kodi to switch to only these `whitelisted` modes.

Some reasons to do this is as follows:
 1. Some people want to switch to a lower resolution when watching 720P content however some people do not want this because when watching live TV the refresh rate switching would happen too often.
 2. We can disable modes that have too high of refresh rate ie. 120HZ.
 3. We can use the exact refresh rate (for low bandwidth systems playing 30hz content at 60hz on a 4k display is difficult)

The problem with this come down to adding yet another setting. More settings generally decrease the user experience. So how can we make things simpler not more complex?
 1. Pre define `whitelisted` modes. (add all refresh rates from the desktop resolution automatically)
 2. Revert to the old behaviour if no `whitelisted` modes are defined.
 3. Hide option at the same settings level as the refresh rate switching option

Currently the logic is this:
 1. for all the methods below remember that it will only select from the whitelist!
 2. the order goes from fuzziest match towards exact match
 3. allow resolutions that are larger than required but have double the refresh rate
 4. allow resolutions that are larger than required but have the correct refresh rate
 5. allow resolutions that are exact and have double the refresh rate
 6. allow resolutions that are exact and have the correct refresh rate

Sample GUI
![Imgur](https://i.imgur.com/1LOp0RO.png)

Sample outputs below.
```
10:35:21.752 T:140499105462656   DEBUG: display width: 1920 vs video width: 1920
10:35:21.752 T:140499105462656   DEBUG: display fps: 60.000000 vs video fps: 29.970030
10:35:21.752 T:140499105462656   DEBUG: float equals: false
10:35:21.752 T:140499105462656   DEBUG: display width: 1920 vs video width: 1920
10:35:21.752 T:140499105462656   DEBUG: display fps: 50.000000 vs video fps: 29.970030
10:35:21.752 T:140499105462656   DEBUG: float equals: false
10:35:21.752 T:140499105462656   DEBUG: display width: 1920 vs video width: 1920
10:35:21.752 T:140499105462656   DEBUG: display fps: 59.940201 vs video fps: 29.970030
10:35:21.752 T:140499105462656   DEBUG: float equals: false
10:35:21.752 T:140499105462656   DEBUG: Matched fuzzy whitelisted Resolution DVI-1: 1920x1080 @ 59.94Hz (19)
10:35:21.752 T:140499105462656   DEBUG: display width: 1920 vs video width: 1920
10:35:21.752 T:140499105462656   DEBUG: display fps: 24.000000 vs video fps: 29.970030
10:35:21.752 T:140499105462656   DEBUG: float equals: false
10:35:21.752 T:140499105462656   DEBUG: display width: 1920 vs video width: 1920
10:35:21.752 T:140499105462656   DEBUG: display fps: 23.976080 vs video fps: 29.970030
10:35:21.752 T:140499105462656   DEBUG: float equals: false
10:35:21.752 T:140499105462656   DEBUG: display width: 1280 vs video width: 1920
10:35:21.752 T:140499105462656   DEBUG: display fps: 60.000000 vs video fps: 29.970030
10:35:21.752 T:140499105462656   DEBUG: float equals: false
10:35:21.752 T:140499105462656  NOTICE: Display resolution ADJUST : DVI-1: 1920x1080 @ 59.94Hz (19) (weight: 0.000)
```
---
```
11:00:03.307 T:140082496567680   DEBUG: display width: 1920 vs video width: 1280
11:00:03.307 T:140082496567680   DEBUG: display fps: 60.000000 vs video fps: 60.000000
11:00:03.307 T:140082496567680   DEBUG: float equals: true
11:00:03.307 T:140082496567680   DEBUG: Matched fuzzy whitelisted Resolution DVI-1: 1920x1080 @ 60.00Hz (17)
11:00:03.307 T:140082496567680   DEBUG: display width: 1920 vs video width: 1280
11:00:03.307 T:140082496567680   DEBUG: display fps: 50.000000 vs video fps: 60.000000
11:00:03.307 T:140082496567680   DEBUG: float equals: false
11:00:03.307 T:140082496567680   DEBUG: display width: 1920 vs video width: 1280
11:00:03.307 T:140082496567680   DEBUG: display fps: 59.940201 vs video fps: 60.000000
11:00:03.307 T:140082496567680   DEBUG: float equals: false
11:00:03.307 T:140082496567680   DEBUG: display width: 1920 vs video width: 1280
11:00:03.307 T:140082496567680   DEBUG: display fps: 24.000000 vs video fps: 60.000000
11:00:03.307 T:140082496567680   DEBUG: float equals: false
11:00:03.307 T:140082496567680   DEBUG: display width: 1920 vs video width: 1280
11:00:03.307 T:140082496567680   DEBUG: display fps: 23.976080 vs video fps: 60.000000
11:00:03.307 T:140082496567680   DEBUG: float equals: false
11:00:03.307 T:140082496567680   DEBUG: display width: 1280 vs video width: 1280
11:00:03.307 T:140082496567680   DEBUG: display fps: 60.000000 vs video fps: 60.000000
11:00:03.307 T:140082496567680   DEBUG: float equals: true
11:00:03.307 T:140082496567680   DEBUG: Matched exact whitelisted Resolution DVI-1: 1280x720 @ 60.00Hz (26)
11:00:03.307 T:140082496567680  NOTICE: Display resolution ADJUST : DVI-1: 1280x720 @ 60.00Hz (26) (weight: 0.000)
```

Please comment to explain why you :heart: or :skull: this idea and any further logic or comments about usability below

Kwiboo commented:
❤️ the idea about using a whitelist but have two comments / suggestions:
1. Modes are currently stored as the index of all resolutions detected, this can cause issues when the user changes hdmi connection, deep color configuration on tv or edid parsing is failing for some reason, when the list of detected resolutions changes the configured whilelist is no longer valid
2. I would like to have some sort of option to default whitelist all 720p/1080p/2160p modes using `appliance.xml` or similar instead of reverting to the old behaviour, on Rockchip I currently have changed the default option for refresh-rate switching to at start/stop using `appliance.xml` and the old behaviour of selecting 4k@60hz for 4k@30hz content is flawed

lrusak commented:
@Kwiboo thanks for the input!
 1. we could possibly hash the the m_resolutions somehow and store that in an unreachable setting or something and validate that and if the hash changes then set the whitelist to default.
 2. I agree, it would be nice to somehow parse for specific modes and allow them to be set by default. This would be huge for applications like your DRMPRIME renderer and on imx6 where there is no HW scaler available to us.

I'll try and think about these two problems

Kwiboo commented:
@lrusak I think that if the setting was saved as a list of strings both my concerns could be addressed.

If the `RESOLUTION_INFO` could be represented by a string, lets say `1920x1080@60hz` to simplify. And the whitelist settings was saved as `1920x1080@23.976hz,1920x1080@60hz`, then it should also be possible to add all 720/1080/2160p modes in `appliance.xml` as the default value.

This will probably require a little bit more complex handling for selecting and saving the whitelisted modes.

lrusak commented:
updated to now save as strings rather than an index.
```
<setting id="videoscreen.whitelist">00192001080060.00000pstd,00192001080050.00000pstd,00192001080059.94020pstd,00192001080024.00000pstd,00192001080023.97608pstd</setting>```

lrusak commented:
pushed another quick fixup that will allow the default behaviour when the whitelist is empty

lrusak commented:
Any other opinions?

FernetMenta commented:
We could add a button to the dialog "Select all/none". Then blacklist fans can select all and then deactivate non desired ones

FernetMenta commented:
We need to persis the whitelisted modes with a platform indenpended name somewhere.

lrusak commented:
> We need to persis the whitelisted modes with a platform indenpended name somewhere.

@FernetMenta can you please explain this statement?

FernetMenta commented:
oops, lots of typos. f*** tablet :)
1)
In this pr the modes don't get persisted yet, right? I see they are read from a setting but I can't find the location where they are stored.
2)
When reading back the setting you use GetResolutionFromString. This naming schema encodes the screen, which is not appropriate for our use case here. If you switch from one display to another, you'll lose the whitelist.

iz8mbw commented:
I am very very interested to that feature of upscaling/downscaling.
Please, is it possible to commit in Kodi 18 this PR so I can test and give a look to it and give my "improvement opionion"? I sue Kodi 18 mainly on Nvidia Shield (Android).
Or give me an Android arm64-v8a APK that I can download and install into my Nvidia Shield.
Thanks.

a1rwulf commented:
@FernetMenta I did some tests. In my guisettings.xml modes are persisted like `<setting id="videoscreen.whitelist">00384002160060.00000pstd,00384002160050.00000pstd,00384002160059.94010pstd,00384002160030.00000pstd,00384002160025.00000pstd,00384002160024.00000pstd,00384002160029.97000pstd,00384002160023.97600pstd,00192001080060.00000pstd,00192001080050.00000pstd,00192001080059.94020pstd,00192001080030.00000pstd,00192001080025.00000pstd,00192001080024.00000pstd,00192001080029.97010pstd,00192001080023.97608pstd</setting>`

I guess https://github.com/xbmc/xbmc/pull/13274/commits/b752a851cfeb5b34c255d3a1c496266255a2de21#diff-8d943449895d0ff6b503e49c8ecf6868R697 is responsible for persisting the modes.

a1rwulf commented:
I also did the following test:
* Hook kodi to my 4k TV
* Enable 4k and 1080p with various resolutions
* Hook kodi to my 1080p monitor
* Only the available 1080p modes are whitelisted
* Hook back to my 4k TV
* All 4k modes and 1080p modes are whitelisted

So from my point of view, everything works as expected.

FernetMenta commented:
Works by chance because in both cases screen number is 0. When finding modes, screen has to be ignored.

a1rwulf commented:
@lrusak can you pick https://github.com/a1rwulf/xbmc/commit/0643374ed19b8557d693a06ccbcc7c9a6c8a0d75 please.
This removes the screen from the serialized modes.

lrusak commented:
thanks @a1rwulf !!

a1rwulf commented:
Anything else needed for this to go in?

lrusak commented:
@a1rwulf I think everyone should be happy now. I should reset the commit dates though.

a1rwulf commented:
If nobody objects I'd like to merge this by end of this week.

iz8mbw commented:
Good! When merged I can test on my Nvidia Shield.

FernetMenta commented on xbmc/guilib/Resolution.cpp:
```diff
@@ -63,16 +67,87 @@ RESOLUTION CResolutionUtils::ChooseBestResolution(float fps, int width, bool is3
-      FindResolutionFromFpsMatch(fps, width, is3D, res, weight);//if that fails use automatic refreshrate selection
+    {
+      if (!FindResolutionFromWhitelist(fps, width, is3D, res)) //find a refreshrate from whitelist
+        FindResolutionFromFpsMatch(fps, width, is3D, res, weight);//if that fails use automatic refreshrate selection
```
why is there a fallback to old behaviour? if a resulution is not whitelisted, it must not be selected

lrusak replied:
You're right, I'll change that. I'll check if the whitelist is empty and return false. Otherwise if there is something in the whitelist it should always return true. 

Does that sound right?

FernetMenta replied:
nope, a res must be whitelisted to get selected. 
- FindResolutionFromWhitelist does not need a return value.
- FindResolutionFromFpsMatch can be deleted

lrusak replied:
No, I'm not going to force people to use the whitelist if they are happy with the current behaviour.

a1rwulf replied:
there’s one compromise that comes into my mind:
if no whitelist is populated yet, add all resolutions to it, so everything would be whitelisted by default.
what do you think?

FernetMenta replied:
>  I'm not going to force people to use the whitelist if they are happy with the current behaviour.

current brahavior will disappear after this change and allowing downscale

FernetMenta replied:
that means: as soon we allow downscale, whitelist has to be used

FernetMenta commented on xbmc/guilib/Resolution.cpp:
```diff
@@ -63,16 +67,87 @@ RESOLUTION CResolutionUtils::ChooseBestResolution(float fps, int width, bool is3
+    // for all the methods below remember that it will only select from the whitelist!
+    // the order goes from fuzziest match towards exact match
+
+    // allow resolutions that are larger than required but have double the refresh rate
```
double fps is wrong. if a user whitelists i.e. 25hz, this should be taken for 25 fps, not 50

lrusak replied:
We've gone through this before.

It will only match resolutions that are in the whitelist, however, it will go through a selection logic to get to the "best" matched resolution.

FernetMenta commented on xbmc/guilib/Resolution.cpp:
```diff
@@ -63,16 +67,87 @@ RESOLUTION CResolutionUtils::ChooseBestResolution(float fps, int width, bool is3
+      resolution = i;
+    }
+
+    // allow resolutions that are larger than required but have the correct refresh rate
```
we only want exact matches, or it won't work

FernetMenta commented:
keep it simple. the idea is to get away from not woring auto select. whitelist means exact match, allow some diff in fractional digits

FernetMenta commented:
Looks like you got the original idea wrong. Whitelisting modes is not an additional feature on top of we currently have. It should be a simplification for too complex not working mode selection we currently have.

lrusak commented:
I don't understand really. This hasn't changed since the original design.

What if I have only 1080p modes selected but I have a 720p24hz file I want to play.

I want it to select the 1080p24hz mode. Not the default 1080p60hz mode.

Most TVs don't have a 720p24hz mode so it would be impossible to select that mode anyways.

The idea was to use whitelist for power users who wanted more control.

FernetMenta commented:
The idea was to simplify complex and not working behaviour. 

> What if I have only 1080p modes selected but I have a 720p24hz file I want to play.

It will use a 1080p. If you want a 720 mode to be selected, whitelist it

> I want it to select the 1080p24hz mode. Not the default 1080p60hz mode.

if 1080p24hz is whitelisted, it will get selected

> Most TVs don't have a 720p24hz mode so it would be impossible to select that mode anyways.

Sorry, I forget on rule. If no exact match is found, allow changing refreshrate but keep resolution

FernetMenta commented:
Look at a very common scenario with just a few modes whitelisted:

1080p60 (desktop), 1080p50, 1080p23.976
2180p23.976

The system should use 2180p23.976 only for content that matches this exactly
All other content searches for a fps match of desktop resolution. Very simple.

If want the system to downscale, they can add 720 or even lower resolution.

lrusak commented:
Ok so I have updated the logic and it will try each method from exact to worst match and return early each time. If someone has an idea how to optimise this code and reduce the amount of duplicate code please let me know.

So now when playing a 720p23.976 video the correct 1080p23.976 mode should be selected.

When playing a 1080p23.976 video the correct 1080p23.976 resolution will be matched right away.

sample log file (verbose logging can be cleaned up)
```
16:34:58.361 T:140077627883072   DEBUG: Trying to find exact refresh rate
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 60.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 0 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 59.940063 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 50.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 24.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 23.976025 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: true
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 60.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 0 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 59.940063 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 50.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: No exact whitelisted resolution matched, trying double refresh rate
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 60.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 0 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 59.940063 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 50.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 24.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 23.976025 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: true
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 60.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 0 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 59.940063 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 50.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: No double refresh rate whitelisted resolution matched, trying larger resolutions
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 60.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 0 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 59.940063 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 50.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 24.000000 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: false
16:34:58.361 T:140077627883072   DEBUG: display width: 1920 vs video width: 1280
16:34:58.361 T:140077627883072   DEBUG: display fps: 23.976025 vs video fps: 23.976025
16:34:58.361 T:140077627883072   DEBUG: display flags: 4 vs video flags: 0
16:34:58.361 T:140077627883072   DEBUG: float equals: true
16:34:58.361 T:140077627883072   DEBUG: Matched fuzzy whitelisted Resolution 1920x1080 @ 23.976025 Hz (24)
16:34:58.361 T:140077627883072  NOTICE: Display resolution ADJUST : 1920x1080 @ 23.976025 Hz (24) (weight: 0.000)
```

FernetMenta commented on xbmc/guilib/Resolution.cpp:
```diff
@@ -63,16 +67,125 @@ RESOLUTION CResolutionUtils::ChooseBestResolution(float fps, int width, bool is3
+
+  for (const auto &mode : indexList)
+  {
+    auto i = CDisplaySettings::GetInstance().GetResFromString(mode.asString());
```
I think we need to test if the resolutions exists in the availalbe modes. User might have connected another display.

FernetMenta commented:
looks good imo. less vorbose logging wold be great.

Memphiz commented:
Didn’t read the commits yet - is the logic where it only switch resolutions on start/stop of a movie touched in this PR? Or does this work like before?

lrusak commented:
@Memphiz that hasn't changed as of yet

3DBuff commented:
I use Raspberry Pi with Kodi as media player and I'm not switching channels with different resolutions and refresh rates. I understand the objective here to limit possible combination of resolution / refresh rates.
I don't really see this as benefit in my case. What I would like to see if you are reworking this section is to add forced set that would ignore HDMI ID tags or lack of them.

In my case if I turn on RPi before turning the TV on I end up with 60Hz only. In order to get the rest of refresh rates (24Hz...) I need to reboot RPi with the TV on. The same scenario is on my projector with the HDMI switch and 7.1 audio extractor on second HDMI out port. If not powered up in correct sequence I end up with 60Hz only.

FernetMenta commented:
> In my case if I turn on RPi before turning the TV on I end up with 60Hz only

this is a shortcoming of RPi and is not related to the work here. RPi should recognize when TV is switched on/off and update the list of available resolutions.

Memphiz commented:
Also Rbpi allows you to persist the edid which would also solve your problem ...

lrusak commented:
Added a fixup because the screen was being ignored so the wrong resolution string was being parsed so I added an extra method to get the current screen and add that to the string.

FernetMenta commented on xbmc/settings/DisplaySettings.cpp:
```diff
@@ -620,6 +620,40 @@ RESOLUTION CDisplaySettings::GetResolutionFromString(const std::string &strResol
   return RES_DESKTOP;
 }
 
+RESOLUTION CDisplaySettings::GetResolutionFromStringIgnoreScreen(const std::string &strResolution)
```
let's take the opportunity and remove screen entirely from resolutions. it makes no sense to load resolution from a screen the app is not running on.

FernetMenta commented:
we can drop screen from resolutions later. that has ever been wrong.

logging shold be less verbose before merge.

apart from this, good to go

garbear commented on xbmc/settings/DisplaySettings.cpp:
```diff
@@ -655,12 +655,17 @@ RESOLUTION CDisplaySettings::GetResolutionForScreen()
 
+static inline bool ModeSort(std::pair<std::string, std::string> i,std::pair<std::string, std::string> j)
+{
+  return (i.second > j.second);
```
Did you intend to sort from least to greatest or greatest to least?

lrusak replied:
This sort greatest to least

lrusak committed:
DisplaySettings: whitelisted mode selection

lrusak committed:
windowing/gbm: rework UpdateResolutions to overwrite RES_DEKSTOP

lrusak commented:
merge?

FernetMenta commented:
+1

iz8mbw commented:
👍

iz8mbw commented:
Why on the latest nightly for Android the mode Whitelist is not available?

sualfred commented:
@lrusak 

This PR breaks multimonitor setups. The switch "Fullscreen on Monitor #i" always jumps back to monitor No.1.

The only way to get the fullscreen on monitor 2 is to switch to window mode, manually move the window to the second monitor, hit alt+enter to get into the fullscreen mode and adjust the resolution to the correct one. On a restart of Kodi I have to do it again.

My monitor setup (if required to know)
Monitor #1 = 2560×1440
Monitor #2 = 1920x1200

Debug log: https://paste.ubuntu.com/p/bqS8q3ZbMK/

FernetMenta commented:
Windows is a bit behind in this regard. Instead of numbering monitors by 1,2,3,.., monitors are supposed to have a name. Screen numbers in resolutions will go away.

natethomas commented:
Hey guys, the Windows devs appear to have only found out about this causing windows breakage a few minutes ago. If a break is found and known, might be worthwhile to ping @afedchin or whoever the platform dev is and let him know at the time, as well as pointing out how and where to fix it. Communication is always good.

FernetMenta commented:
I am wondering if there is any dev left who has turned on github notifications.

afedchin commented:
> I am wondering if there is any dev left who has turned on github notifications.

I'm wondering if you have a time to inspect all PRs. It's easy to skip something really important (like personal ping) in this spam from GH. Even If I had notifications enabled this PR has the title which makes no words what it breaks multimon configs but in fact it does.

FernetMenta commented:
Since this PR does affect all platforms I assumed that all platform devs did observe it. Seems I was wrong.
No bad sttitude. Sorry.

FernetMenta commented on xbmc/settings/DisplaySettings.cpp:
```diff
@@ -656,10 +656,39 @@ RESOLUTION CDisplaySettings::GetResolutionForScreen()
+  RESOLUTION res = CDisplaySettings::GetInstance().GetDisplayResolution();
+  RESOLUTION_INFO info = CDisplaySettings::GetInstance().GetResolutionInfo(res);
+
+  for (auto index = (unsigned int)RES_DESKTOP; index < CDisplaySettings::GetInstance().ResolutionInfoSize(); ++index)
```
modes start at RES_CUSTOM

lrusak replied:
I had it that way initially but had complaints that two of the same resolutions would be shown

lrusak replied:
I had it that way initially but had complaints that two of the same resolutions would be shown

FernetMenta replied:
so you removed the double entry by increasing the list? magic :)

lrusak replied:
I misremembered, it was that some destoop resolution werensh't own in the whitelist. But we've figured out why on slack 😝

the-dreamer commented:
Whitelist broke my 1080i resolution. 1080p ist not possible on my tv. In Resolution i can choose 1080i but it switch to a 1080p 50/59 or 60 resolution.

FernetMenta commented:
> 1080p ist not possible on my tv

time to replace your tv with a newer model

the-dreamer commented:
an Oled  >65inch is on my whishlist... until then is there a way to disable it logic... until end of april it was no issue.

bylderup commented:
Whitelist broke my 1080i resolution, too.
I tested this now with 2 different fresh untouched Installations on SD Cards with Mecool M8Spro+ 
and three different TVs - and I won´t buy new TVs..

In Settings-System-Display-Resolution 1920x1080i is listed.
If I try to activate this resolution then nothing happens!
As a follow up no 1920x1080i options are listed in whitelist.

Only way to get Display Resolution to 1920x1080i is manually editing the guisettings.xml

Didn´t have this before whitelist was introduced in KODI
[log-2018-06-10-11.37.24.zip](https://github.com/xbmc/xbmc/files/2087852/log-2018-06-10-11.37.24.zip)

fritsch commented:
Guys. Kodi outputs progressive. Whatever you force on your 1080 interlaced  resolutions makes a bsolutely no sense.

bylderup commented:
just tested with LibreELEC-S905.arm-9.0-nightly-20180607-36c2619
same problem
I am not a developer - for whatever reason 2 of my TVs do have the best picture quality with 1920x1080i 50 Hz.  This resolution is listed in Kodi settings and doesn´t work anymore since whitelist feature was introduced. And thats a bug imhop. No 1080 interlaced neither on a Hdready nor on a full HD TV.

fritsch commented:
@lrusak could you workout with your LE guys ;-) about pending / in place AMLogic hacks? Not sure how in their universe progressive content looks fine on interlaced outpud modes - therefore no idea what those builds include.

bylderup commented:
Don´t know if we are talking about the same thing.  All I said is : 
since whitelist was introduced in Kodi, it is not possible so set the Display Resolution in Kodi to 1920x1080i.

# cat /sys/class/amhdmitx/amhdmitx0/disp_cap
480p60hz
576p50hz
720p60hz
1080i60hz*
720p50hz
1080i50hz

the-dreamer commented:
I need to relativate my initial problem. After update to a more recent Version Kodi from 2nd June my system switched to SD resolution. i tried to set to via resolution 1080i which was showed me twice. both was resulting in "mode not supported" on my Plasma HdReady screen. I downgraded and sleep a night. Today i tried again. Same bahaviour today. No suprise. But now i was looking into xrandr and it switched always to 60Hz 1080p when i selected 1080i. After connecting to system via vncviewer i tried different option and after swithing to 1080p i found 25p for refresh rate. I know 25p is the equivalent to 50i. Now i have at least picture again with 1080i. I am not sure if it was desired like it is now. For frame rate switching i have used advancedsetting xml to avoid switching to 50p for example.

lrusak commented:
Pleas use the forum's and/or trac for issue reports. This is a dev space.
